File: calculator/src/test/java/com/turlir/calculator/converter/ValueTest.java
package com.turlir.calculator.converter;

import org.junit.Test;
import org.mockito.Mockito;

import java.math.BigDecimal;
import java.math.MathContext;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Locale;

import static org.junit.Assert.assertEquals;

public class ValueTest {

    private final DecimalFormat mFormat;

    public ValueTest() {
        mFormat = new DecimalFormat();
        DecimalFormatSymbols settings = new DecimalFormatSymbols(new Locale("ru"));
        mFormat.setDecimalFormatSymbols(settings);
        mFormat.setMaximumFractionDigits(340); // see doc DecimalFormat
        mFormat.setMinimumIntegerDigits(1);
    }

    @Test
    public void findFormatPath() {
        String t = "1234.1234567890123456789";
        BigDecimal bd = new BigDecimal(t, MathContext.UNLIMITED);

        DecimalFormatSymbols fs = new DecimalFormatSymbols(new Locale("ru"));
        DecimalFormat df = new DecimalFormat();
        df.setDecimalFormatSymbols(fs);
        df.setMaximumFractionDigits(340);

        String format = df.format(bd);
        assertEquals("1 234,1234567890123456789", format);
    }

    @Test
    public void displayTest() {
        Value v = new Value("1234.123456", false);
        Visual view = v.view();
        Printer printer = Mockito.spy(new SimplePrinter(mFormat));
        view.print(printer);
        assertEquals("1 234,123456", printer.toString());
    }

    @Test
    public void lastSymbolSeparatorTest() {
        Value v = new Value("1234", true);
        Visual view = v.view();
        Printer printer = Mockito.spy(new SimplePrinter(mFormat));
        view.print(printer);
        assertEquals("1 234,", printer.toString());
    }

    private class SimplePrinter implements Printer {

        private final StringBuilder mBuilder;
        private final DecimalFormat mFormat;
        private final String mSeparator;

        SimplePrinter(DecimalFormat format) {
            mFormat = format;
            mSeparator = String.valueOf(mFormat.getDecimalFormatSymbols().getDecimalSeparator());
            mBuilder = new StringBuilder();
        }

        @Override
        public void append(String part) {
            mBuilder.append(part);
        }

        @Override
        public void append(BigDecimal part) {
            String str = mFormat.format(part);
            append(str);
        }

        @Override
        public void appendSeparator() {
            append(mSeparator);
        }

        @Override
        public int length() {
            return 0;
        }

        @Override
        public void reset() {
            throw new UnsupportedOperationException();
        }

        @Override
        public String toString() {
            return mBuilder.toString();
        }
    }
}

File: calculator/src/test/java/com/turlir/calculator/converter/ValueTest.kt
package com.turlir.calculator.converter

import com.turlir.calculator.member.Value
import org.junit.Assert.assertEquals
import org.junit.Test
import java.math.BigDecimal
import java.math.MathContext
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.*

class ValueTest {

    private val mFormat = DecimalFormat()

    init {
        val settings = DecimalFormatSymbols(Locale("ru"))
        mFormat.decimalFormatSymbols = settings
        mFormat.maximumFractionDigits = 340 // see doc DecimalFormat
        mFormat.minimumIntegerDigits = 1
    }

    @Test
    fun findFormatPath() {
        val t = "1234.1234567890123456789"
        val bd = BigDecimal(t, MathContext.UNLIMITED)
        val format = mFormat.format(bd)
        assertEquals("1 234,1234567890123456789", format)
    }

    @Test
    fun displayTest() {
        val v = Value("1234.123456", false)
        val view = v.view()
        val printer = SimplePrinter(mFormat)
        view.print(printer)
        assertEquals("1 234,123456", printer.toString())
    }

    @Test
    fun lastSymbolSeparatorTest() {
        val v = Value("1234", true)
        val view = v.view()
        val printer = SimplePrinter(mFormat)
        view.print(printer)
        assertEquals("1 234,", printer.toString())
    }

    private inner class SimplePrinter constructor(private val mFormat: DecimalFormat) : Printer {

        private val mBuilder = StringBuilder()
        private val mSeparator: String = mFormat.decimalFormatSymbols.decimalSeparator.toString()

        override fun append(part: String) {
            mBuilder.append(part)
        }

        override fun append(part: BigDecimal) {
            val str = mFormat.format(part)
            append(str)
        }

        override fun appendSeparator() {
            append(mSeparator)
        }

        override fun length() = 0

        override fun reset() = throw UnsupportedOperationException()

        override fun toString() = mBuilder.toString()
    }
}